### PR TITLE
Deploy command

### PR DESCRIPTION
**Pull request by OleksandrRoshchuk**

### CODE REVIEW

File: src/commands/deploy.js [hunk 1]
```diff
@@ -47,7 +66,7 @@ async function setupDeployKeyIDFile(keyID, { stdin, stdout, stderr }, config) {
     switch (e.code) {
       case 'ENOENT':
       case 'ENOTDIR':
-        return mkdir(dir, { mode: 0x755 }).catch((e) => {
+        return mkdir(dir, { mode: 0o755 }).catch((e) => {
```
Comment (gabesullice) [inline src/commands/deploy.js:75]: Good catch.

File: src/commands/deploy.js [hunk 2]
```diff
@@ -4,7 +4,17 @@ import { setupDeployKey } from "./deploy-key.js";
...
+ *
+ * @param args
+ * @param stdin
+ * @param stdout
+ * @param stderr
+ * @param {Config} config
+ */
 export default async function deploy(args, { stdin, stdout, stderr }, config) {
   let key = config.deployKey;
   if (key === undefined) {
```
Comment (gabesullice) [inline src/commands/deploy.js:20]: Did you encounter this problem or is this defensive?

File: src/lib/api.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    const createFrom = await getLinkData(key, createFormLink, { config });
+    const testDir = fs.mkdtempSync(`${tmpdir()}${sep}`);
+    await printWhile(
+        stderr,
+        {
+            pending: "Creating deploy file…",
+            resolved: "Deploy file creation complete.\n",
+            rejected: "Deploy file creation failed.\n",
+        },
+        zip(deployFolder, `${testDir}${sep}deploy.zip`),
```
Comment (gabesullice) [inline src/lib/api.js]: :+1:  for `${sep}`

File: src/main.js [hunk 4]
```diff
@@ -48,6 +48,7 @@ export default async function main(cwd, argv, stdStreams, env) {
       await handle(e, commandName, argv, stdStreams, context);
     } else {
       const config = Config.read(context);
+      console.log(config);
```
Comment (gabesullice) [inline src/main.js]: Stray `console.log`

File: src/lib/api.js [hunk 5]
```diff
@@ -0,0 +1,92 @@
+import {DeployKey} from "./deploy-key.js";
+import {zip} from "zip-a-folder";
+import fs from "fs";
+import {UnknownError} from "./errors.js";
+import parse from "./resource.js";
+import { tmpdir } from "node:os";
+import {sep} from "node:path";
```
Comment (gabesullice) [inline src/lib/api.js]: Can you `npm run prettier`?

File: src/lib/api.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import {printWhile} from "./cli.js";
+
+export async function apiDeploy(key, deployFolder, releaseNote, {stdin, stdout, stderr}, config) {
+    const project = await DeployKey.getProject(key, { config });
+    const releaseOverviewLink = project.links.get('releases').href;
+    let releaseOverview = await getLinkData(key, releaseOverviewLink, { config });
+    const createFormLink = releaseOverview.links.get('create-form').href;
+    const createFrom = await getLinkData(key, createFormLink, { config });
+    const testDir = fs.mkdtempSync(`${tmpdir()}${sep}`);
```
Comment (gabesullice) [inline src/lib/api.js]: nit: can you import `mkdtempSync` directly instead of `fs` just to be consistent.

File: src/lib/api.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
+import {DeployKey} from "./deploy-key.js";
+import {zip} from "zip-a-folder";
+import fs from "fs";
+import {UnknownError} from "./errors.js";
+import parse from "./resource.js";
+import { tmpdir } from "node:os";
+import {sep} from "node:path";
+import {printWhile} from "./cli.js";
+
+export async function apiDeploy(key, deployFolder, releaseNote, {stdin, stdout, stderr}, config) {
```
Comment (gabesullice) [inline src/lib/api.js]: I think this is fine as-is for this commit, but passing in `stdin` _et al._ pierces the abstraction a little by coupling this to the command line, which may not be as useful when reusing this for the GitHub action. Perhaps we need a function to build the zip and another function to upload it, then the `printWhile`s could be in the `deploy` command.

**Review by gabesullice — CHANGES_REQUESTED** [review]
This is looking really good. Just a few required changes, then a suggestion about breaking `apiDeploy` into 2 commands, but you can choose to do that in a separate PR (or not at all if you disagree).

File: src/commands/deploy.js (hunk 2 above)
Comment (OleksandrRoshchuk) [reply src/commands/deploy.js:20]: Just to be consistent with the deploy-key.js

File: src/commands/deploy.js (hunk 1 above)
Comment (OleksandrRoshchuk) [reply src/commands/deploy.js:75]: Thanks

File: src/lib/api.js (hunk 5 above)
Comment (OleksandrRoshchuk) [reply src/lib/api.js]: Done

File: src/main.js (hunk 4 above)
Comment (OleksandrRoshchuk) [reply src/main.js]: done

File: src/lib/api.js (hunk 6 above)
Comment (OleksandrRoshchuk) [reply src/lib/api.js]: Done

File: src/lib/api.js (hunk 7 above)
Comment (OleksandrRoshchuk) [reply src/lib/api.js]: Decided to postpone this part until have more info regarding github actions

**Review by gabesullice — APPROVED** [review]

**Pull request merged**